### ISSUE TITLE
[Bug] native attention backend: garbage/empty output on Qwen3-8B + O(batch^2) decode slowness & OOM (TP4, v6e-4)

### ISSUE THREAD

**Issue by Rodrian7** [sgl-project/sglang-jax#1297]
## Summary

On TPU `v6e-4` (TP4), the **`native` attention backend** produces **incorrect generations** (garbage token-salad or empty/degraded output) for `Qwen/Qwen3-8B`, while the `fa` backend on the *exact same* model and server config (only `--attention-backend` changed) is fully correct. Separately, the `native` backend is **pathologically slow** at decode (~0.7 tok/s per request) and **OOMs** at non-trivial context lengths.

Together these block the `qwen3-8b-native` accuracy gate planned in #1200 §1.1.1: the case cannot run a meaningful end-to-end eval within any reasonable time budget, and even when it runs, output correctness is unstable.

---

## Environment

- **Repo:** `sgl-project/sglang-jax` @ commit `07c0bed` (branch `worktree-nightly-4tpu-accuracy`)
- **Python** 3.12.12 · **jax / jaxlib** 0.8.1 · **sgl_jax** 0.0.2
- **Hardware:** TPU `v6e-4` (4 chips, ~32 GB HBM/chip)
- **Model:** `Qwen/Qwen3-8B`, `bf16`, TP4
- `JAX_COMPILATION_CACHE_DIR` warm; model under `--download-dir /dev/shm/`

### Server launch (only `--attention-backend` + the memory knobs differ)

```bash
# native (under test)
python -m sgl_jax.launch_server --trust-remote-code --model-path Qwen/Qwen3-8B \
  --host 127.0.0.1 --port 30011 --tp-size 4 --dp-size 1 --nnodes 1 --node-rank 0 \
  --dist-init-addr 127.0.0.1:35011 --skip-server-warmup --random-seed 3 \
  --download-dir /dev/shm/ --dtype bfloat16 --page-size 256 \
  --attention-backend native \
  --mem-fraction-static 0.6 --max-prefill-tokens 4096 --context-length 3072 \
  --precompile-token-paddings 4096 --precompile-bs-paddings 52 --max-running-requests 52 \
  --device tpu

# fa (baseline)
python -m sgl_jax.launch_server --trust-remote-code --model-path Qwen/Qwen3-8B \
  --host 127.0.0.1 --port 30011 --tp-size 4 --dp-size 1 --nnodes 1 --node-rank 0 \
  --dist-init-addr 127.0.0.1:35011 --skip-server-warmup --random-seed 3 \
  --download-dir /dev/shm/ --dtype bfloat16 --page-size 256 \
  --attention-backend fa \
  --mem-fraction-static 0.8 --max-prefill-tokens 8192 --max-running-requests 32 \
  --precompile-token-paddings 8192 --precompile-bs-paddings 32 \
  --device tpu
```

### Cases

The first **8 GSM8K test examples**, obtained via `eval.simple_eval_gsm8k.get_examples(8)`
(the canonical `openai/grade-school-math` `test.jsonl`). Targets (in order):
`[18, 3, 70000, 540, 20, 64, 260, 160]`. Greedy decode (`temperature=0.0`), `max_tokens=2048`,
`chat_template_kwargs={"enable_thinking": false}` — **identical for both backends**.

### Repro script (`cmp_probe.py`, run once per backend against the live server)

```python
import time, sys, json, concurrent.futures as cf
from openai import OpenAI
from eval.simple_eval_gsm8k import get_examples, INSTRUCTION_TEMPLATE, parse_answer, score_gsm8k, extract_answer_from_target

tag = sys.argv[1]  # "fa" or "native"
N = 8
client = OpenAI(base_url="http://127.0.0.1:30011/v1", api_key="x")
exs = get_examples(N)
extra = {"chat_template_kwargs": {"enable_thinking": False}}

def ask(i, ex):
    r = client.chat.completions.create(
        model="Qwen/Qwen3-8B",
        messages=[{"role": "user", "content": INSTRUCTION_TEMPLATE.format(input=ex["question"])}],
        temperature=0.0, max_tokens=2048, extra_body=extra)
    txt = r.choices[0].message.content or ""
    return dict(i=i, tok=r.usage.completion_tokens, ans=parse_answer(txt),
                tgt=extract_answer_from_target(ex["answer"]), text=txt)

rows = []
with cf.ThreadPoolExecutor(max_workers=N) as pool:
    for f in cf.as_completed([pool.submit(ask, i, e) for i, e in enumerate(exs)]):
        d = f.result(); d["ok"] = score_gsm8k(d["tgt"], d["ans"]); rows.append(d)
rows.sort(key=lambda x: x["i"])
print(tag, "correct=%d/%d" % (sum(r["ok"] for r in rows), N))
```

---

### Result

| backend | result |
|---|---|
| **fa** | **8/8 correct**, every answer coherent step-by-step |
| **native** | **mixed** — garbage, empty, degraded, and correct all appear under identical greedy decode |

Per-question contrast (same question index, same config, **only** `--attention-backend` differs):

| Q | target | native | fa |
|---|---|---|---|
| Q0 | 18 | `127` tok, ans **18 ✓**, but degraded opening: `"To the   Janet’s ducks lay **16 eggs per day**. ..."` | `256` tok, ans **18 ✓**, clean |
| Q1 | 3 | **garbage**: `"To No tri很方便 or)-(aN  Equation日凌晨iaz占比做起 (XHR frm�nim两级rax(ls扣inox_EXTRAEH hiệp \"-HCIMul ..."` → ans **2 ✗** | `109` tok, ans **3 ✓** |
| Q3 | 540 | **empty** (`completion_tokens=1`, no answer) → **✗** ¹ | `172` tok, ans **540 ✓** |
| Q5 | 64 | `188` tok, ans **64 ✓** | `189` tok, ans **64 ✓** |

¹ Q3’s empty output was reproduced on a clean server in a prior 16-example run; in the 8-example run the same garbage/degradation pattern recurs on other indices. The `native` run is slow enough that not all 8 completed before it was stopped, but the failures above are decisive.

`fa` produces 8/8 correct, fully coherent answers (sample below); `native` produces token salad / empty / degraded openings on the same prompts. This isolates the defect to the `native` backend (not the model, chat template, or harness).

<details><summary>fa Q3 (correct, for contrast)</summary>

```
To solve the problem, we need to calculate the total number of meters James runs in a week.
### Step 1: Determine the number of sprints per week
James runs 3 sprints 3 times a week.  => 3 × 3 = 9 sprints
### Step 2: distance per sprint = 60 meters
### Step 3: 9 × 60 = 540 meters
Answer: 540
```
</details>

Consistent with this, an earlier `native` MMLU-64 run scored **0.03** (≈ random).

### Hypothesis (unverified)

The degenerate output may come from a numerical edge case in the decode mask + softmax path: a fully-masked row (all `-inf`) through `max`/`exp`/`sum` (`native_backend.py:299-310`) can yield `NaN`, poisoning sampling. `_apply_decode_mask` (`native_backend.py:389-420`) only fills the mask for `:num_seqs` rows, leaving padded rows fully masked. **This is a hypothesis** — the exact mechanism has not been isolated.

---

### Root cause (from `native_backend.py`)

The decode path concatenates **all running sequences' KV** into one dense score tensor, then masks off-diagonal blocks:

- `native_backend.py:254` — `attn_logits = jnp.einsum("qhd,khd->qhk", q_heads, k_heads)`, where the `k` axis spans **all running sequences' KV concatenated** (`key_len ≈ max_running × seq_len`).
- `native_backend.py:389-420` (`_apply_decode_mask`) — a **block-diagonal** mask keeps, per query row, only its own sequence's keys; the other `(running-1)/running` of the dot-products are computed then discarded.

For `running=52`, every query attends over all 52 sequences' keys and ~**98 % (51/52)** of the computed scores are masked away. The full `[q_len, heads, key_len]` score tensor is also **materialized** (no flash-style tiling).

### Measured impact (native, context 3072)

- **Decode step ≈ 1.45 s/step** (`decode_log_interval=40` → ~58 s between decode logs ÷ 40). Constant for `running=8` and `running=52`, because decode is compiled to `precompile_bs_paddings=[52]`.
- **Per-request ≈ 0.7 tok/s**; aggregate = `0.7 × running` (≈ 36 tok/s at 52). Wall-clock is bounded by the longest single output, not total tokens.
- A 188-token answer ⇒ ~**273 s** of decode vs ~**2 s** prefill (decode dominates ~130×). For comparison, `fa` answered all 8 questions in 1–3 s each.
- A thinking-on answer (~1500 tokens) ⇒ ~36 min for a single question; full GSM8K never completes in a practical budget.
- **Precompile**: native EXTEND 137 s + DECODE 94 s ≈ **231 s**, vs fa 12 s + 6 s ≈ **18 s**.

### OOM is the score tensor, **not** the KV cache

With default `--context-length 40960` and `--max-running-requests 16`, the score tensor is
`bf16[max_prefill_tokens=4096, kv_heads=8, context×running=655360] ≈ 43 GB` → OOM (HBM ~32 GB/chip).
The KV cache is *not* the constraint: `max_total_num_tokens = 434,944` (Fused KV 59.77 GB across TP4), peak usage ~6 %. Only `context 3072 × max_running 52` (score K = 159,744) fits; `max_running 56` already OOMs.

---

## Impact

- **Blocks** the `qwen3-8b-native` GSM8K gate in #1200 §1.1.1. (The case as checked in also omits `--context-length`, so it OOMs on the default 40960 before it can run.)
- The reference/`native` attention backend currently cannot be validated end-to-end on this model.

## Suggested directions (non-prescriptive)

1. **Fix the correctness defect first** — isolate the garbage/empty output (masked-row softmax `NaN`?), and add a small **`native` vs `fa` numerical-equivalence unit test** (same q/k/v → `allclose`), the standard way reference attention backends are validated (SGLang `torch_native`, vLLM, HF eager, llama.cpp). sglang-jax currently has no kernel-level attention-equivalence test.
2. **Performance** — avoid the O(batch²) cross-sequence recompute (per-sequence / paged attention rather than one dense block-diagonal matmul) and avoid materializing the full score tensor.
3. For the nightly gate (#1200 §1.1.1) — until the above land, gate `native` via the numerical-equivalence unit test rather than an end-to-end GSM8K/MMLU eval.

---

cc — context from #1200 (Nightly §1.1). Reproduced manually on `v6e-4`; full per-question dumps available on request.

**Comment by Rodrian7** [comment]
## Update: `batch=1` isolation — the correctness bug is in the **multi-sequence / batched decode path**, and the perf root-cause needs a correction

Re-ran the same 8 GSM8K examples (same greedy / `enable_thinking=False` config) on `native` with **`--max-running-requests 1 --precompile-bs-paddings 1`** (true batch=1, requests sent sequentially). Everything else identical to the original repro.

### Correctness: failures disappear at batch=1

| Q | native @ `max_running=52` | native @ `max_running=1` | fa (baseline) |
|---|---|---|---|
| Q0 | 127 tok, degraded opening (`"To the  Janet's ducks..."`), ans 18 | **257 tok, clean, ans 18 ✓** | 256 tok, ans 18 ✓ |
| Q1 | **garbage** token-salad, ans ✗ | **109 tok, clean, ans 3 ✓** | 109 tok, ans 3 ✓ |
| Q2 | — | **302 tok, clean, ans 70000 ✓** | 299 tok, ans 70000 ✓ |
| Q3 | **empty** (`completion_tokens=1`), ✗ | **142 tok, clean, ans 540 ✓** | 172 tok, ans 540 ✓ |

At batch=1, every previously-failing question (garbage Q1, empty Q3, degraded Q0) becomes **clean and correct**, and the outputs closely match `fa` (length and answer). The single-sequence attention math is therefore fine; **the defect is in the batched / multi-sequence decode path** — i.e. the block-diagonal mask / cross-sequence KV gather that only engages when `running > 1` (`_apply_decode_mask`, `native_backend.py:389-420`; cross-sequence concat at `:254`). This narrows the earlier "NaN from a fully-masked row" hypothesis to the multi-sequence masking specifically.

### Correction to Problem 2 (performance root cause)

The original write-up attributed the slowness primarily to the O(batch²) cross-sequence recompute. The batch=1 measurement shows that is **not** the dominant term:

| config | per-step | per-request rate |
|---|---|---|
| `max_running=1` | ~0.80 s/step | **1.25 tok/s** |
| `max_running=52` | ~1.45 s/step | 0.69 tok/s (×52 = ~36 tok/s aggregate) |

If the O(batch²) block-diagonal waste were dominant, batch=1 should be ~52× faster per token; instead it is only ~1.8× faster. So:

- There is a **large fixed ~0.8 s/step cost even at batch=1** — a single request already decodes at only ~1.25 tok/s. This (not the batch² term) is the primary slowness; a 257-token answer still takes ~205 s at batch=1.
- The cross-sequence O(batch²) term is real but secondary — it only grows the step from ~0.8 s to ~1.45 s going from 1→52 sequences.

The OOM analysis (score tensor, not KV cache) and the precompile cost (~215 s, dominated by the 4096-token EXTEND compile, ~independent of bs) are unchanged.

### Implications for next steps

1. **Correctness fix** should target the multi-sequence decode path (batched mask / cross-sequence handling), since batch=1 is already correct.
2. The suggested **`native` vs `fa` equivalence unit test** must cover the **batched (multi-sequence) case**, not just a single sequence — a single-sequence test would pass and miss this bug entirely.
3. **Performance** needs both: (a) reduce the fixed per-step cost that already makes batch=1 slow (~0.8 s/step), and (b) the per-sequence/paged attention to remove the batch² term.

Config for this run (only the two flags changed vs the original native repro):
```
--attention-backend native --mem-fraction-static 0.6 --max-prefill-tokens 4096 \
--context-length 3072 --precompile-token-paddings 4096 \
--precompile-bs-paddings 1 --max-running-requests 1
```

**Comment by real3stone** [comment]
Thanks for the detailed `batch=1` isolation. I'd like to take the **correctness** half of this (leaving the O(batch²) / fixed per-step perf as a separate follow-up).

Root cause on current `main`:

- Decode `cache_loc` is built **page-aligned per request** — each sequence occupies `ceil(seq_len/page_size)*page_size` slots, with offsets `= cumsum(aligned_lens)` (`ScheduleBatch._merge_cache_loc`, `schedule_batch.py` ~L2396).
- But `_apply_decode_mask` reconstructs the per-sequence key ranges with a **dense** `cumsum(seq_lengths)` (`native_backend.py` ~L415).

The two disagree by exactly the page padding, so each sequence's mask window slides off its real keys. The drift for sequence `i` is the total page padding of the sequences ahead of it, so it accumulates; the failure condition is precisely **"some sequence other than the last has a length that is not a multiple of `page_size`"**, which is the normal case in serving. The displaced window picks up a neighbour's page-tail slots — those hold real allocated slot indices (allocator page ids start at 1), so `is_valid = loc > 0` doesn't remove them — and drops some of the sequence's own keys, including the most recent one.

`batch=1` is immune (single block at offset 0), which matches your isolation. And with `--page-size 256` over short prompts the drift exceeds the sequence length entirely: reproducing two ~100-token sequences gives a drift of 156 and an overlap of **0/100**, i.e. the attention output is built on completely unrelated KV — that's why it's garbage rather than merely degraded.

Minimal repro, no TPU and no model needed: `_apply_decode_mask` with `page_size=4`, `seq_lens=[3, 2]` selects `[3, 5)` for the 2nd sequence, whose real keys are at `[4, 6)`. A sharper pair: `seq_lens=[4, 3]` is correct while `[3, 4]` is not — same lengths, same total padding, only the order differs.

A few notes from digging into it:

- `_merge_cache_loc` carries the comment *"padding slots are never read on-device: attention kernels (RPA v3 / MLA v2 / native) bound page reads by cu_kv_lens / seq_lens"*, and `test_merge_cache_loc.py` deliberately excludes page tails from its assertions. So the page-aligned layout is a stated mechanism `native` was named as honouring, and this is a violation of it rather than an ambiguity. (That comment is justifying buffer reuse and reasons about the buffer's unwritten tail — that part is fine, the displaced windows stay inside the freshly written region. It's the `seq_lens` bound itself that isn't honoured.)
- `_apply_extend_mask` has the same dense assumption (`k_starts = cumsum(seq_lens) - seq_lens`) and I've confirmed it reproduces identically for batched prefill. I'd keep that as a separate change since that mask derives block-diagonality from `cumsum`-based batch ids, which assumes contiguous blocks — but flagging it so it doesn't look like the backend is fully fixed.
- The same window reconstruction is also unaware of the **fixed per-rank stride** used when `cache_loc` / `seq_lens` are merged across DP ranks, so under `dp_size > 1` every rank ≥ 1 sequence gets a shifted window even with page alignment handled (pre-existing — the dense cumsum was equally wrong under DP; your TP4 repro has `data=1` so it doesn't hit this). I'd treat that as its own change too, on its merits: it's a stride problem orthogonal to page alignment, and the right fix has to decide where the mask learns the per-rank segmentation and to pin down (with `dp > 1` mask tests — straightforward with a forced host device count) the merge-side stride invariants it would then rely on. That's a self-contained unit of review, not a bigger version of this one.

Proposed scope (partially addresses #1297): make `_apply_decode_mask` page-aware, and add a batched native-vs-reference decode equivalence unit test (CPU, `page_size > 1` + unequal-length sequences, MHA/GQA/MQA, bucket padding) registered in `unit-test-cpu` — there's currently no unit test for this backend at all. The performance work (fixed per-step cost + O(batch²)), the extend-path mask, and the DP-stride handling would be separate follow-ups.

Does this scope look right, and is the issue free to take?
